lib/IRGen/MetadataRequest.cpp

```diff
@@ -444,14 +444,29 @@ static llvm::Value *emitObjCMetadataRef(IRGenFunction &IGF,
   return emitObjCMetadataRefForMetadata(IGF, classPtr);
 }
 
+static bool isObjCGenericParam(GenericTypeParamDecl *param) {
+  if (!param) return false;
+  auto nominal = param->getDeclContext()->getSelfNominalTypeDecl();
+
+  if (!nominal) return false;
+  return nominal->isTypeErasedGenericClass();
+}
+
 // Get the type that exists at runtime to represent a compile-time type.
 CanType IRGenModule::getRuntimeReifiedType(CanType type) {
-  // Leave type-erased ObjC generics with their generic arguments unbound, since
-  // the arguments do not exist at runtime.
+  // The generic parameters of ObjC generic classes don't exist at runtime, so
+  // we want to represent them using types that do exist at runtime.
   return CanType(type.transform([&](Type t) -> Type {
+    // Leave type-erased ObjC generics with their generic arguments unbound.
     if (CanType(t).isTypeErasedGenericClassType()) {
       return t->getAnyNominal()->getDeclaredType()->getCanonicalType();
     }
+    // Erase ObjC generic parameters to an equivalent existential.
+    if (auto archT = t->getAs<PrimaryArchetypeType>()) {
+      if (isObjCGenericParam(archT->getInterfaceType()->getDecl())) {
```

Comment:
I'd like to get rid of the backpointer to the GenericTypeParamDecl from GenericTypeParamType one day since it causes problems all over the place. It's not the end of the world to add a new usage here, but is there any other way you can think of expressing this?

Reply:
I think we'd need a bit on `GenericTypeParamType` saying that it's type-erased. I could add an `isTypeErased()` accessor to it that's currently implemented in terms of `getDecl()`, but could be changed to work from a flag bit in the future.

Reply:
(The only other approach I could come up with was to make the ObjC-generic-ness a feature of the `GenericSignature` and therefore known from the `PrimaryArchetypeType`, but that seems like it'd wind through a ton of code and I don't see a good way to fit objc-generic-ness into the existing model.)

Reply:
can we look at the nominal type whose metadata we're instantiating? Ie, if you have `Foo<T>`, you're erasing `T` to some protocol or class iff `Foo` is an Objective-C generic, correct?

Reply:
... and if Foo _is not_ an Objective-C lightweight generic class, then erasing the generic parameter to its existential bound is not valid in general, because existentials do not self-conform.

Reply:
We don't have `Foo<T>` here; if we did, the existing case (converting it to an unbound type for `Foo`) would cover it. We have something like `T` or `T.Type` here, typically because we wrote an extension method with a `T` or `T.Type` parameter. So what I'm trying to handle is a type like this:

```
(metatype_type thick
  (primary_archetype_type address=0x7fdf2d595f18 class layout=AnyObject name=ObjectType))
```

I haven't been able to come up with a way to go from the `PrimaryArchetypeType` to the declaration the generic parameter belongs to without going through `getInterfaceType()->getDecl()`, and I don't really have any other context to work from in this method. Maybe if I moved `getRuntimeReifiedType()` to IRGenFunction, I'd be able to see if I'm in a pseudogeneric function. But isn't this really more a property of the parameter or the generic signature used in the archetype? Surely generic parameters for other types also pass through this function.

Reply:
What happens if you have a canonical type (getDecl() is null)?

Reply:
Checking at a higher level if we're inside of a pseudogeneric function sounds like a good idea. Maybe emitArchetypeTypeMetadataRef() is the right place for this? Also would this subsume the existing erasure to UnboundGenericType? It would be nice to eliminate that as well.